source/common/http/http2/codec_impl.cc

```diff
@@ -157,6 +157,18 @@ void ConnectionImpl::StreamImpl::destroy() {
   parent_.stats_.pending_send_bytes_.sub(pending_send_data_->length());
 }
 
+ConnectionImpl::ServerStreamImpl::~ServerStreamImpl() {
+  // Only the downstream stream should clear the downstream of the
+  // memory account.
+  //
+  // There are cases where a corresponding upstream stream dtor might
+  // be called, but the downstream stream isn't going to terminate soon
+  // such as StreamDecoderFilterCallbacks::recreateStream().
+  if (buffer_memory_account_) {
+    buffer_memory_account_->clearDownstream();
```

Comment:
Doing this from the destructor may be too late and result in calls to resetStream after deferred deletion started.  Consider hooking into `StreamImpl::destroy` which is called before scheduling the deferred deletion.